### PR TITLE
Allow users to whitelist hosts from the option page

### LINKED ISSUE #17434: Add ability to disable HTTPS Everywhere for specific domains without first visiting the site
Type: feature request

Hi, I've run into some issues using HTTPS Everywhere while on sites on my corporate networks, or on sites that are "public" but are proxied via Microsoft Application Proxy to our internal network. I'm not sure why this is creating problems (in this case it was our JIRA instance), but in trying to get around it, I've noticed that I can only disable HTTPS Everywhere for a site if I am on the site and disable it through the extensions right click menu.

It would be great if I could explicitly whitelist domains that HTTPS Everywhere would be disabled for without having to do it for every site as needed (in my case, I'd disable it for my corporate domains). I think it would make sense to add this functionality into the extension settings, maybe even under "Advanced".

Thanks for your consideration!

### LINKED ISSUE #18503: Allow end user to add *.neverssl.com into the disabled sites list
Type: query

I find that ``neverssl.com`` is unusable on EASE mode, even when its domain is added to my "HTTPS Everywhere Sites Disabled" list, as it currently redirects you to a url like ``lxzmwvtdsnrhkfbc.neverssl.com`` (where part before the dot is random data) to fix problems with stale in-browser cache

Add-ons_Manager

So I like a way to add ``*.neverssl.com`` into the disabled sites list so that ``neverssl.com`` would work for me on EASE mode

I have written a custom ruleset for ``neverssl.com`` and added it into firefox via debugging rulesets but it still doesn't work with EASS

Debugging_Rulesets

It would be the good if ``*.neverssl.com`` can be excluded in https-everywhere. If not, please allow the user to add it themselves via a ruleset, or adding disabled sites with wildcard in the extension's preference UI, or both, so that ``neverssl.com`` continues to work for users that choose to use EASS mode

### PR DESCRIPTION
**Pull request by cschanaj**
This PR will:

- Add UI to add disabled site without require a first visit (close #17434)
- Add wildcard support for disabled site (close #18503)
- Allow us to get rid of the double-negation toggle (follow-up ~#18565~ #18590)

Note that:

- ~this PR have temporarily removed the unit test for the `nullIterable` object because some refactoring works are blocked by #18563. I will submit a PR to restore the unit test once that is merged.~ See 8a6bace

- wildcard matching in this PR is not really working with WASM enabled because of #18589. So the review of this PR will be much easier with the fix in https://github.com/EFForg/https-everywhere-lib-core/pull/1 merged. 

P.S. Personally, I hope this can be merged before a release is made (#18618) if it is not too urgent.

**EDIT:** Updated the incorrect PR reference.

### CODE REVIEW

File: chromium/background-scripts/util.js [hunk 1]
```diff
@@ -110,8 +110,8 @@ function getWildcardExpressions(host) {
 
   // now eat away from the left, with *, so that for x.y.z.google.com we
   // check *.y.z.google.com, *.z.google.com and *.google.com
-  for (let i = 1; i <= segmented.length - 2; i++) {
-    const tmp = "*." + segmented.slice(i, segmented.length).join(".");
+  for (let i = 1; i < segmented.length - 1; i++) {
+    const tmp = ["*", ...segmented.slice(i, segmented.length)].join(".");
```
Comment (pipboy96) [inline chromium/background-scripts/util.js]: Good idea, makes more sense IMO.

**Comment by cschanaj** [comment]
@zoracon would you take a look at this PR as well?

**Comment by zoracon** [comment]
Reviewing today

**Comment by pipboy96** [comment]
I think it would also make sense to allow the user to remove temporary exclusions created by "Open insecure page for this session only" option on EASE error page.

File: chromium/pages/options/ux.js [hunk 2]
```diff
@@ -331,6 +327,18 @@ document.addEventListener("DOMContentLoaded", () => {
     }
   });
 
+  const add_disabled_site = document.getElementById("add-disabled-site");
+  const disabled_site_input = document.getElementById("disabled-site");
+  disabled_site_input.setAttribute("placeholder", chrome.i18n.getMessage("options_enterDisabledSite"));
+
+  add_disabled_site.addEventListener("click", function() {
+    if (disabled_site_input.value !== "") {
```
Comment (zoracon) [inline chromium/pages/options/ux.js]: Could you also add some validation here as well?

**Comment by cschanaj** [comment]
> wildcard matching in this PR is not really working with WASM enabled because of #18589. So the review of this PR will be much easier with the fix in EFForg/https-everywhere-lib-core#1 merged.

I am so sorry that I was wrong on this statement. It seems there is uncommited changes on my local branch and the wildcard matching is not working because the refactoring work was partially done.

**Comment by cschanaj** [comment]
@zoracon @pipboy96 this is ready for a second round of reviews. thanks.

**Comment by cschanaj** [comment]
ping @zoracon for a second review. thanks.

**Comment by zoracon** [comment]
Pulling this down and testing today. Apologies for the delay

**Comment by pipboy96** [comment]
@zoracon I think we should also allow adding whitelisted domains/disabled rulesets and maybe force/disable EASE from managed storage (enterprise policy).

**Comment by zoracon** [comment]
Pulling this down and testing today. Apologies for the delay.

**Pull request merged**